src/JustEat.StatsD/EndpointLookups/CachedIpEndpointSource.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Net;
+
+namespace JustEat.StatsD.EndpointLookups
+{
+    /// <summary>
+    /// cache the IPEndPoint and only go to the source when it expires
+    /// </summary>
+    public class CachedIpEndpointSource : IPEndPointSource
+    {
+        private IPEndPoint _cachedValue;
+        private DateTime _expiry;
+        private readonly int _cacheDurationSeconds;
+
+        private readonly IPEndPointSource _inner;
+
+        public CachedIpEndpointSource(IPEndPointSource inner, int cacheDurationSeconds)
```

Comment:
Personally I'd prefer a TimeSpan so calling code looks a bit more readable:

```csharp
new CachedIpEndpointSource(inner, 5 * 60);
```

vs.

```csharp
new CachedIpEndpointSource(inner, TimeSpan.FromMinutes(5));
```

It also allows the consumer arbitrary precision (though why you'd cache for less than one second I don't know).